### DISCUSSION TITLE
Equivalent of `cat ~/{.bashrc,.bash_history}` in nushell

### DISCUSSION THREAD

**Discussion by atahabaki** [nushell/nushell#9256] [Q&A] [answered]
In bash, we can shorten our command by:

```bash
git add ~/.config/nushell/{config.nu,env.nu}
```

Is there any equivalent syntax in nushell?

**Comment by amtoine** [comment]
the simplest i know is the following :yum: 
```bash
[config.nu env.nu] | each {|file|
    git add ([~/.config/nushell $file] | path join)
}
```

**Reply by atahabaki** [reply]
I know looping over it is the way to go, but bash does it for me.

**Reply by fdncred** [reply]
You could play around with the `glob` command and see if it works for you. I just did this and it lists the files.
```
glob nushell/{config.nu,env.nu}
```

**Reply by atahabaki** [reply]
@fdncred still not like bash though, how could you use with git? for loop?

**Reply by fdncred** [reply]
I'd probably use the `each` command like
```
glob nushell/{config.nu,env.nu} | each {|r| do_something $r}
```

**Reply by atahabaki** [reply]
@fdncred it does not work if the file/folder is not available. For example, you can't create a file like:

```
glob ~/.config/nvim/lua/{core,plugin} | each {|p| mkdir -p $p}
```

`glob` works, only if the file or the folder already exists. That's a shame, but this shell is just at the stage of beginning, i hope someday it'll be better than bash.

The best approach, isp is still writing a for loop...

**Reply by fdncred** [reply]
ya, `glob` and `ls` are both looking for files or folders, if they don't exist it won't work. The only alternative to non-existing files is to have a list like amtoine suggested above.

**Reply by ghost** [reply]
you can pass a array to an external command and `git add` allows adding multiple files at once -> `git add (glob ~/.config/nvim/lua/{core,plugin})` should work and is a bit shorter than a `each` loop.

**Reply by atahabaki** [reply]
@fdncred still not equivalent, if the file or folder is not existing, you can't for example combine with `touch` like bash do:

```
touch /path/to/sth/{init,config}.rc
```

**Reply by fdncred** [reply, marked as answer]
@atahabaki That's cool. I think the way nushell uses globs can be greatly improved, there are several things that I don't like about it too. However, I'm not sure we'll ever approach bash's brevity, it's kind of a non-goal.

**Comment by Taywee** [comment]
In my use, this works:

```shell
git add ([config.nu env.nu] | each {[$env.HOME .config nushell $in] | path join})
# or
git add ([config.nu env.nu] | each {$'($env.HOME)/.config/nushell/($in)'})
```

I prefer this to the `each` approach with `git add` inside, because this one will still do it all in one `git add` execution.

If you're worried about command line argument length limit, you can defer to xargs as usual:

```shell
[config.nu env.nu] | each {$'($env.HOME)/.config/nushell/($in)'} | str join "\u{0}" | xargs -0 git add
```

Personally, I like this better than the Bash `{}`, because I can use filter and interpolate arbitrarily, rather than just splatting the whole list (though that is useful behavior, and the Bash equivalent is more concise when that's all you want).

**Reply by atahabaki** [reply]
For me, this is killing the simplicity of `git add (~/.config/{nushell/{config,env}.nu,starship.toml} | str expand)`
If you prefer, more advanced approach, of course an iteration is better. But look at the example, how simple is that?

**Reply by Taywee** [reply]
That's fair enough.  Given your example of piping into the hypothetical `str expand`, it seems like something that would be easy enough to add as a plugin, [given how simple the functionality really is](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html).

**Comment by Taywee** [comment]
@atahabaki 
I actually went and [wrote a plugin for this functionality](https://crates.io/crates/nu-plugin-bexpand), and a [backing crate](https://crates.io/crates/bexpand) for it.  It's pretty functional and robust.  I was wrong about "easy", though.  I got caught in a handful of weeds getting this thing working the way I wanted, but I'm pretty happy with it.

**Reply by Taywee** [reply]
Check it out:

```nushell
> '~/.config/nushell/{config.nu,env.nu}' | str bexpand
╭───┬─────────────────────────────╮
│ 0 │ ~/.config/nushell/config.nu │
│ 1 │ ~/.config/nushell/env.nu    │
╰───┴─────────────────────────────╯

> '~/.config/nushell/{config.nu,env.nu}' | str bexpand | path expand
╭───┬──────────────────────────────────────╮
│ 0 │ /home/home/.config/nushell/config.nu │
│ 1 │ /home/home/.config/nushell/env.nu    │
╰───┴──────────────────────────────────────╯

> [a{🤐..🤒}b c{🧛..🦜..31}d e{15..-23..8}f] | str bexpand
╭────┬───────╮
│  0 │ a🤐b  │
│  1 │ a🤑b  │
│  2 │ a🤒b  │
│  3 │ c🧛d  │
│  4 │ c🦼d  │
│  5 │ c🦝d  │
│  6 │ e15f  │
│  7 │ e7f   │
│  8 │ e-1f  │
│  9 │ e-9f  │
│ 10 │ e-17f │
╰────┴───────╯

> ['{🤐..🤖}' '{🤤..🥔..32}' "{\u{1F995}..\u{1F990}}"] | str bexpand
╭────┬────╮
│  0 │ 🤐 │
│  1 │ 🤑 │
│  2 │ 🤒 │
│  3 │ 🤓 │
│  4 │ 🤔 │
│  5 │ 🤕 │
│  6 │ 🤖 │
│  7 │ 🤤 │
│  8 │ 🥄 │
│  9 │ 🦕 │
│ 10 │ 🦔 │
│ 11 │ 🦓 │
│ 12 │ 🦒 │
│ 13 │ 🦑 │
│ 14 │ 🦐 │
╰────┴────╯
```

Pretty nifty, if I do say so myself.

**Reply by fdncred** [reply]
Cool! Thanks going in awesome-nu https://github.com/nushell/awesome-nu !

Read through the crate and the plugin. Very nice work!

**Reply by Taywee** [reply]
Thanks, I'm happy to have it there!

**Reply by atahabaki** [reply]
@Taywee it's looking great. On the "easy" part, I've thought the same. I faced really cool problems though but I managed to overcome them all. Anyway  stay tuned for #9290.

**Reply by Taywee** [reply]
I didn't even realize you had already made one of your own!  I should really use the search functionality before spending hours reinventing the wheel :laughing: 

Your work looks good, but I think it's still missing some functionality of the Bash equivalent.  Doesn't look like you can do `f{a,}`. I've used that occasionally for things like `cp file.sh{,.backup}`.

Mine's not complete either.  It's missing the ability to specify `{08..10}` to get `['08', '09', '10']`, but I didn't want to implement that the same as Bash's, because you can't specify the fixed width with negatives.  There's no way of specifying equal width for `{-1..10}`, expanding to `['-1', '00', '01'...`, because a leading zero doesn't work when both ends are already the length you want without it, and having a leading zero for eg. `{-01..100}` is ugly and less clear.

I was thinking of a syntax like `{=-1..10}`, with a leading equal sign to mean "make all numbers padded to the same width.  It looks weird with negatives (`{=-1..1000..300}` would be `['-001', '0299', '0599', '0899']`), but that'll be weird regardless.

**Reply by Taywee** [reply]
I've just added that ability to bexpand:

```nushell
> '{=-1..1000..300}' | str bexpand
╭───┬──────╮
│ 0 │ -001 │
│ 1 │ 0299 │
│ 2 │ 0599 │
│ 3 │ 0899 │
╰───┴──────╯

> '{=-1..0001000..300}' | str bexpand
╭───┬─────────╮
│ 0 │ -000001 │
│ 1 │ 0000299 │
│ 2 │ 0000599 │
│ 3 │ 0000899 │
╰───┴─────────╯

> '{-001..0001000..300}' | str bexpand
╭───┬─────╮
│ 0 │ -1  │
│ 1 │ 299 │
│ 2 │ 599 │
│ 3 │ 899 │
╰───┴─────╯
```

**Reply by atahabaki** [reply]
i think '_' would be better suited. 

`{_3.._5}` -> `[03,04,05]`

I'll add {a..z} and a couple more features to bracoxide crate.

`f{,a}` is basic. I didn't know it was useful. priority 0 for me, first feature to implement to v0.2.0.

Your plugin looks awesome BTW.

**Reply by Taywee** [reply]
I didn't want a number prefix that counts as part of the length, because then you couldn't tell it to make `{-1..10}` zero-padded, because both endpoints are already the full length.  It works fine for non-negative numbers, but the negative thing bothers me.  So I just thought "`=` means 'make the widths equal'" and that's good enough for me.  Really, zero-padded negatives probably aren't all that useful anyway.

> Your plugin looks awesome BTW.

Thanks!  I'm pretty happy with it.  Pleasantly surprised by how easy it is to write nushell plugins, so I'm definitely going to exercise that a lot more.

Your PR looks good.  I hope it gets in!  Your error messages are definitely better than mine (I'm just spitting out nom's error messages directly).  I'll probably continue to maintain my plugin anyway, because it was fun to write and I like working with nom.
